File: AocCsharp/Day01/Day01.cs
using AocCsharp.Utils;

namespace AocCsharp.Day01;

public static class Day01
{
    public static void Run()
    {
        var testInputPath = Path.Combine("..", "inputs", "day01_test.txt");
        var inputPath = Path.Combine("..", "inputs", "day01.txt");

        // Part 1: Count how many times the dial points at 0
        TestRunner.RunSolution(
            "Part 1",
            Part1,
            testInputPath,
            inputPath,
            expectedTestResult: 3
        );

        // Part 2: To be unlocked after completing Part 1
        TestRunner.RunSolution(
            "Part 2",
            Part2,
            testInputPath,
            inputPath,
            expectedTestResult: 6
        );
    }

    public static int Part1(string input)
    {
        // Parse the rotations (handle spaces and newlines)
        var rotations = input.Split(new[] { ' ', '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries);
        
        int position = 50;  // Starting position
        int zeroCount = 0;

        foreach (var rotation in rotations)
        {
            var direction = rotation[0];
            var distance = int.Parse(rotation.Substring(1));

            if (direction == 'L')
            {
                position = (position - distance) % 100;
                if (position < 0) position += 100;
            }
            else if (direction == 'R')
            {
                position = (position + distance) % 100;
            }

            if (position == 0)
            {
                zeroCount++;
            }
        }

        return zeroCount;
    }

    public static int Part2(string input)
    {
        // Parse the rotations (handle spaces and newlines)
        var rotations = input.Split(new[] { ' ', '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries);
        
        int position = 50;  // Starting position
        int zeroCount = 0;

        foreach (var rotation in rotations)
        {
            var direction = rotation[0];
            var distance = int.Parse(rotation.Substring(1));

            if (direction == 'R')
            {
                // Moving right: count multiples of 100 in range (position, position + distance]
                // Equivalent to floor((pos + dist) / 100)
                zeroCount += (position + distance) / 100;
                position = (position + distance) % 100;
            }
            else if (direction == 'L')
            {
                // Moving left: count multiples of 100 in range [position - distance, position)
                // We use floor division logic. 
                // Count = floor((pos - 1) / 100) - floor((pos - dist - 1) / 100)
                
                // Since pos is always [0, 99], floor((pos - 1) / 100) is:
                // 0 if pos > 0
                // -1 if pos == 0
                int startFloor = (position - 1) < 0 ? -1 : 0;
                int endFloor = (int)Math.Floor((double)(position - distance - 1) / 100);
                
                zeroCount += startFloor - endFloor;
                
                position = (position - distance) % 100;
                if (position < 0) position += 100;
            }
        }

        return zeroCount;
    }
}

File: AocC/day01.c
#include "day01.h"
#include "runner.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

void day01_run(void) {
    const char* test_path = "../inputs/day01_test.txt";
    const char* real_path = "../inputs/day01.txt";

    run_solution("Part 1", day01_part1, test_path, real_path, 3);
    run_solution("Part 2", day01_part2, test_path, real_path, 6);
}

int64_t day01_part1(const char* input) {
    int position = 50;  // Starting position
    int zero_count = 0;

    const char* ptr = input;
    while (*ptr) {
        // Skip whitespace
        while (*ptr && isspace((unsigned char)*ptr)) ptr++;
        if (!*ptr) break;

        char direction = *ptr++;
        int distance = 0;

        // Parse number
        while (*ptr && isdigit((unsigned char)*ptr)) {
            distance = distance * 10 + (*ptr - '0');
            ptr++;
        }

        if (direction == 'L') {
            position = (position - distance) % 100;
            if (position < 0) position += 100;
        } else if (direction == 'R') {
            position = (position + distance) % 100;
        }

        if (position == 0) {
            zero_count++;
        }
    }

    return zero_count;
}

int64_t day01_part2(const char* input) {
    int position = 50;  // Starting position
    int zero_count = 0;

    const char* ptr = input;
    while (*ptr) {
        // Skip whitespace
        while (*ptr && isspace((unsigned char)*ptr)) ptr++;
        if (!*ptr) break;

        char direction = *ptr++;
        int distance = 0;

        // Parse number
        while (*ptr && isdigit((unsigned char)*ptr)) {
            distance = distance * 10 + (*ptr - '0');
            ptr++;
        }

        if (direction == 'R') {
            // Moving right: count multiples of 100 in range (position, position + distance]
            zero_count += (position + distance) / 100;
            position = (position + distance) % 100;
        } else if (direction == 'L') {
            // Moving left: count multiples of 100 in range [position - distance, position)
            int start_floor = (position - 1) < 0 ? -1 : 0;

            int temp = position - distance - 1;
            int end_floor;
            if (temp >= 0) {
                end_floor = temp / 100;
            } else {
                // Floor division for negative numbers
                end_floor = (temp - 99) / 100;
            }

            zero_count += start_floor - end_floor;

            position = (position - distance) % 100;
            if (position < 0) position += 100;
        }
    }

    return zero_count;
}
